mithril-common/src/chain_reader/pallas_chain_reader.rs

```diff
@@ -0,0 +1,349 @@
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::fs;
+
+    use pallas_network::{
+        facades::NodeServer,
+        miniprotocols::{
+            chainsync::{BlockContent, Tip},
+            Point,
+        },
+    };
+    use tokio::net::UnixListener;
+
+    use super::*;
+
+    use crate::test_utils::TempDir;
+
+    /// Enum representing the action to be performed by the server.
+    enum ServerAction {
+        RollBackward,
+        RollForward,
+    }
+
+    /// Enum representing whether the node has agency or not.
+    #[derive(Debug, PartialEq)]
+    enum HasAgency {
+        Yes,
+        No,
+    }
+
+    /// Returns a fake specific point for testing purposes.
+    fn get_fake_specific_point() -> Point {
+        Point::Specific(
+            1654413,
+            hex::decode("7de1f036df5a133ce68a82877d14354d0ba6de7625ab918e75f3e2ecb29771c2")
+                .unwrap(),
+        )
+    }
+
+    /// Returns a fake chain point for testing purposes.
+    fn get_fake_chain_point_backwards() -> ChainPoint {
+        ChainPoint::from(get_fake_specific_point())
+    }
+
+    /// Returns a fake chain point for testing purposes.
+    fn get_fake_chain_point_forwards() -> ChainPoint {
+        Tip(get_fake_specific_point(), 1337).into()
+    }
+
+    /// Creates a new work directory in the system's temporary folder.
+    fn create_temp_dir(folder_name: &str) -> PathBuf {
+        TempDir::create_with_short_path("pallas_chain_observer_test", folder_name)
+    }
+
+    fn get_fake_raw_block() -> Vec<u8> {
+        let raw_block = include_str!("../../../mithril-test-lab/test_data/blocks/shelley1.block");
+
+        hex::decode(raw_block).unwrap()
+    }
+
+    fn get_fake_scanned_block() -> ScannedBlock {
+        let raw_block = get_fake_raw_block();
+        let multi_era_block = MultiEraBlock::decode(&raw_block).unwrap();
+
+        ScannedBlock::convert(multi_era_block, 0)
+    }
+
+    /// Sets up a mock server for related tests.
+    ///
+    /// Use the `action` parameter to specify the action to be performed by the server.
+    async fn setup_server(
+        socket_path: PathBuf,
+        action: ServerAction,
+        has_agency: HasAgency,
+    ) -> tokio::task::JoinHandle<()> {
+        tokio::spawn({
+            async move {
+                if socket_path.exists() {
+                    fs::remove_file(&socket_path).expect("Previous socket removal failed");
+                }
+
+                let known_point = get_fake_specific_point();
+                let unix_listener = UnixListener::bind(socket_path.as_path()).unwrap();
+                let mut server = NodeServer::accept(&unix_listener, 10).await.unwrap();
+
+                let chainsync_server = server.chainsync();
+
+                chainsync_server.recv_while_idle().await.unwrap();
+
+                chainsync_server
+                    .send_intersect_found(known_point.clone(), Tip(known_point.clone(), 1337))
```

Comment:
Is it make sense to create a constant for 1337 ?

Reply:
Done